src/matrixrtc/CallMembership.ts

```diff
@@ -19,19 +19,85 @@ import { deepCompare } from "../utils";
+    Array.isArray(data.foci_preferred);
+
+const checkSessionsMembershipData = (data: SessionMembershipData): void => {
+    const prefix = "Malformed session membership event: ";
+    if (typeof data.device_id !== "string") throw new Error(prefix + "device_id must be string");
+    if (typeof data.call_id !== "string") throw new Error(prefix + "call_id must be string");
+    if (typeof data.application !== "string") throw new Error(prefix + "application must be a string");
+    if (typeof data.foci_active?.type !== "string") throw new Error(prefix + "foci_active.type must be a string");
+    if (Array.isArray(data.foci_preferred)) throw new Error(prefix + "foci_preferred must be an array");
+    // optional elements
+    if (data.created_ts && typeof data.created_ts !== "number") throw new Error(prefix + "created_ts must be number");
+
+    // application specific data (we first need to check if they exist)
+    if (data.scope && typeof data.scope !== "string") throw new Error(prefix + "scope must be string");
+};
+
+// Legacy session membership data
+export interface CallMembershipDataLegacy {
+    application: string;
     call_id: string;
     scope: CallScope;
     device_id: string;
+    membershipID: string;
     created_ts?: number;
     expires?: number;
     expires_ts?: number;
     foci_active?: Focus[];
-    membershipID: string;
 }
+const isLegacyCallMembershipData = (data: CallMembershipData): data is CallMembershipDataLegacy =>
+    "membershipID" in data && "foci_active" in data && Array.isArray(data.foci_active);
+const checkCallMembershipDataLegacy = (data: CallMembershipDataLegacy): void => {
+    const prefix = "Malformed legacy rtc membership event: ";
+    if (!(data.expires || data.expires_ts)) {
+        throw new Error(prefix + "expires_ts or expires must be present");
+    }
+    if (data.expires) {
+        if (typeof data.expires !== "number") {
+            throw new Error(prefix + "expires must be numeric");
+        }
+    }
+    if (data.expires_ts) {
+        if (typeof data.expires_ts !== "number") {
+            throw new Error(prefix + "expires_ts must be numeric");
+        }
+    }
+
+    if (typeof data.device_id !== "string") throw new Error(prefix + "device_id must be string");
+    if (typeof data.call_id !== "string") throw new Error(prefix + "call_id must be string");
+    if (typeof data.application !== "string") throw new Error(prefix + "application must be a string");
+    if (typeof data.membershipID !== "string") throw new Error(prefix + "membershipID must be a string");
+    // optional elements
+    if (data.created_ts && typeof data.created_ts !== "number") throw new Error(prefix + "created_ts must be number");
+    // application specific data (we first need to check if they exist)
+    if (data.scope && typeof data.scope !== "string") throw new Error(prefix + "scope must be string");
+};
+
+export type CallMembershipData = CallMembershipDataLegacy | SessionMembershipData;
```

Comment:
How about using an XOR?
```suggestion
export type CallMembershipData = XOR<CallMembershipDataLegacy, SessionMembershipData>;
```

This will let type checking do a lot more with `CallMembershipData` and may let you remove may of the runtime typechecks altogether (as long as typehints can be trusted / there aren't insidious `any`s lying around).